list.c

```diff
@@ -0,0 +1,62 @@
+#include <stdlib.h>
+#include <stdint.h>
+#include "list.h"
+
+// simple linked list implementation
+
+struct list *list_new(void)
+{
+    return calloc(1, sizeof(struct list));
+}
+
+void list_clear(struct list *list)
+{
+    struct list_item *next, *cur = list->head;
+    while (cur != NULL) {
+        next = cur->next;
+        free(cur);
+        cur = next;
+    }
+}
+
+void *list_append(struct list *list, void *data)
+{
+    struct list_item *item = malloc(sizeof(struct list_item));
+    if (item == NULL) {
+        return NULL;
+    }
+    item->prev = list->tail;
+    item->next = NULL;
+    item->data = data;
+    if (list->head == NULL) {
+        list->head = item;
+    }
+    if (list->tail != NULL) {
+        list->tail->next = item;
+    }
+    return item;
+}
+
+void list_remove(struct list *list, void *data)
+{
+    struct list_item *next, *cur;
+    // is list empty?
+    if (list->head == NULL) {
+        return;
+    }
+    cur = list->head;
+    while (cur != NULL) {
+        next = cur->next;
+        if (cur->data == data) {
+            if (cur->prev != NULL) {
+                cur->prev->next = cur->next;
+            }
+            if (cur->next != NULL) {
+                cur->next->prev = cur->prev;
+            }
+            free(cur);
+            break;
+        }
+        cur = next;
+    }
```

Comment:
dont you consider that "head" and "tail" can be modified in this function, when you remove a list item?